functional test failure on master: sign_up - sign up with a verified account forces the user to sign in

× firefox on any platform - sign_up - sign up with a verified account forces the user to sign in (1.206s)
AssertionError: expected '' to equal 'signin0.9366139306221157@restmail.net'
  at Function.assert.equal  <node_modules/intern/node_modules/chai/chai.js:2028:10>
  at Command.<anonymous>  <tests/functional/sign_up.js:317:24>
  at runCallback  <node_modules/leadfoot/Command.js:526:31>
  at Command.<anonymous>  <node_modules/leadfoot/Command.js:543:11>
  at <node_modules/dojo/Promise.ts:393:15>
  at run  <node_modules/dojo/Promise.ts:237:7>
  at <node_modules/dojo/nextTick.ts:44:3>
  at doNTCallback0  <node.js:419:9>
  at process._tickCallback  <node.js:348:13>
  at Command.then  <node_modules/leadfoot/Command.js:542:10>
  at <tests/functional/sign_up.js:315:16>
  at onSettled  <app/bower_components/fxa-js-client/fxa-client.js:1312:9>
  at Immediate.flush [as _onImmediate]  <app/bower_components/fxa-js-client/fxa-client.js:1019:6>
  at processImmediate [as _immediateCallback]  <timers.js:383:17>

--- a/tests/functional/change_password.js
+++ b/tests/functional/change_password.js
@@ -40,7 +40,7 @@ define([
         return FunctionalHelpers.fillOutChangePassword(context, FIRST_PASSWORD, SECOND_PASSWORD);
       })
 
-      .then(FunctionalHelpers.visibleByQSA('a[href="/confirm_account_unlock"]'))
+      .then(FunctionalHelpers.visibleByQSAErrorHeight('a[href="/confirm_account_unlock"]'))
 
       .findByCssSelector('a[href="/confirm_account_unlock"]')
         .click()

--- a/tests/functional/delete_account.js
+++ b/tests/functional/delete_account.js
@@ -40,7 +40,7 @@ define([
         return FunctionalHelpers.fillOutDeleteAccount(context, PASSWORD);
       })
 
-      .then(FunctionalHelpers.visibleByQSA('#delete-account .error'))
+      .then(FunctionalHelpers.visibleByQSAErrorHeight('#delete-account .error a[href="/confirm_account_unlock"]'))
       .end()
 
       .findByCssSelector('a[href="/confirm_account_unlock"]')

--- a/tests/functional/lib/helpers.js
+++ b/tests/functional/lib/helpers.js
@@ -176,19 +176,28 @@ define([
    *
    * @param {String} selector
    *        QSA compatible selector string
+   * @param {Object} options
+   *        options include polling `timeout`, element `offsetWidth` and element `offsetHeight`
    */
-  function visibleByQSA(selector, timeout) {
-    timeout = timeout || 10000;
+  function visibleByQSA(selector, options) {
+    options = options || {};
+    var timeout = options.timeout || 10000;
 
-    return pollUntil(function (selector) {
+    return pollUntil(function (selector, options) {
+      var offsetHeight = options.offsetHeight || 0;
+      var offsetWidth = options.offsetWidth || 0;
       var match = document.querySelectorAll(selector);
 
       if (match.length > 1) {
         throw new Error('Multiple elements matched. Make a more precise selector');
       }
 
-      return match[0] && match[0].offsetWidth > 0 ? true : null;
-    }, [ selector ], timeout);
+      return match[0] && (match[0].offsetWidth > offsetWidth && match[0].offsetHeight > offsetHeight) ? true : null;
+    }, [ selector, options ], timeout);
+  }
+
+  function visibleByQSAErrorHeight(selector) {
+    return visibleByQSA(selector, {offsetHeight: 10});
   }
 
   function noSuchElement(context, selector) {
@@ -964,6 +973,7 @@ define([
     testUrlPathnameEquals: testUrlPathnameEquals,
     thenify: thenify,
     type: type,
-    visibleByQSA: visibleByQSA
+    visibleByQSA: visibleByQSA,
+    visibleByQSAErrorHeight: visibleByQSAErrorHeight
   };
 });

--- a/tests/functional/oauth_sign_in.js
+++ b/tests/functional/oauth_sign_in.js
@@ -39,7 +39,7 @@ define([
         return FunctionalHelpers.fillOutSignIn(context, email, PASSWORD);
       })
 
-      .then(FunctionalHelpers.visibleByQSA('.error'))
+      .then(FunctionalHelpers.visibleByQSAErrorHeight('.error a[href="/confirm_account_unlock"]'))
       .end()
 
       .findByCssSelector('a[href="/confirm_account_unlock"]')

--- a/tests/functional/reset_password.js
+++ b/tests/functional/reset_password.js
@@ -681,7 +681,7 @@ define([
     'open /reset_password page, enter unknown email, wait for error': function () {
       return fillOutResetPassword(this, email)
         // The error area shows a link to /signup
-        .then(FunctionalHelpers.visibleByQSA('.error a[href="/signup"]'))
+        .then(FunctionalHelpers.visibleByQSAErrorHeight('.error a[href="/signup"]'))
         .findByCssSelector('.error a[href="/signup"]')
           .click()
         .end()

--- a/tests/functional/sign_in.js
+++ b/tests/functional/sign_in.js
@@ -113,7 +113,7 @@ define([
       var email = TestHelpers.createEmail();
       return fillOutSignIn(this, email, PASSWORD)
         // The error area shows a link to /signup
-        .then(FunctionalHelpers.visibleByQSA('.error a[href="/signup"]'))
+        .then(FunctionalHelpers.visibleByQSAErrorHeight('.error a[href="/signup"]'))
         .findByCssSelector('.error a[href="/signup"]')
           .click()
         .end()

